File: world/scene.py
"""
Scene Transition Manager v3
───────────────────────────
Smooth horizontal slide transitions between environments.
Memories persist across transitions.
"""
import time
import numpy as np
import cv2
from config import settings
from world.environments import EnvironmentManager, Environment


class Scene:
    def __init__(self, env_manager: EnvironmentManager):
        self.env_manager = env_manager
        self.transitioning = False
        self.transition_start_time = 0.0
        self.transition_duration = settings.ENV_TRANSITION_DURATION
        self.transition_direction = 0
        self.from_env: Environment = None
        self.to_env: Environment = None

    def trigger_swipe(self, direction: int):
        """
        direction = -1  ->  previous environment (left hand triple-swipe)
        direction = +1  ->  next environment (right hand triple-swipe)
        """
        if self.transitioning:
            return
        self.from_env = self.env_manager.current()
        if direction > 0:
            self.to_env = self.env_manager.next()
        else:
            self.to_env = self.env_manager.previous()
        self.transition_direction = direction
        self.transition_start_time = time.time()
        self.transitioning = True

    def get_background(self) -> np.ndarray:
        """Returns current background, interpolated during transitions."""
        if not self.transitioning:
            return self.env_manager.current().image.copy()

        elapsed = time.time() - self.transition_start_time
        progress = min(1.0, elapsed / self.transition_duration)

        if progress >= 1.0:
            self.transitioning = False
            return self.env_manager.current().image.copy()

        # Ease-out cubic
        ease = 1.0 - (1.0 - progress) ** 3
        h, w = self.from_env.image.shape[:2]
        shift = int(ease * w)

        canvas = np.zeros_like(self.from_env.image)
        if self.transition_direction > 0:
            # Slide right: old exits right, new enters from left
            if shift < w:
                canvas[:, shift:] = self.from_env.image[:, :w - shift]
            if shift > 0:
                canvas[:, :shift] = self.to_env.image[:, w - shift:]
        else:
            # Slide left: old exits left, new enters from right
            if shift < w:
                canvas[:, :w - shift] = self.from_env.image[:, shift:]
            if shift > 0:
                canvas[:, w - shift:] = self.to_env.image[:, :shift]

        return canvas

    def current_environment(self) -> Environment:
        return self.env_manager.current()

    def is_original_room(self) -> bool:
        return self.env_manager.is_original_room() and not self.transitioning

File: src/world/scene.rs
// Scene Transition Manager
// Matches world/scene.py exactly.

use std::time::Instant;
use image::{ImageBuffer, Rgb};
use crate::config;
use crate::world::environments::{Environment, EnvironmentManager};

pub struct Scene {
    pub env_manager: EnvironmentManager,
    pub transitioning: bool,
    pub transition_start_time: Instant,
    pub transition_duration: f32,
    pub transition_direction: i32,
    from_image: ImageBuffer<Rgb<u8>, Vec<u8>>,
    to_image: ImageBuffer<Rgb<u8>, Vec<u8>>,
}

impl Scene {
    pub fn new(env_manager: EnvironmentManager) -> Self {
        let cur_img = env_manager.current().image.clone();
        Self {
            env_manager,
            transitioning: false,
            transition_start_time: Instant::now(),
            transition_duration: config::ENV_TRANSITION_DURATION,
            transition_direction: 0,
            from_image: cur_img.clone(),
            to_image: cur_img,
        }
    }

    pub fn trigger_swipe(&mut self, direction: i32) {
        if self.transitioning {
            return;
        }
        self.from_image = self.env_manager.current().image.clone();
        if direction > 0 {
            self.to_image = self.env_manager.next().image.clone();
        } else {
            self.to_image = self.env_manager.previous().image.clone();
        }
        self.transition_direction = direction;
        self.transition_start_time = Instant::now();
        self.transitioning = true;
    }

    pub fn get_background(&mut self) -> ImageBuffer<Rgb<u8>, Vec<u8>> {
        if !self.transitioning {
            return self.env_manager.current().image.clone();
        }

        let elapsed = self.transition_start_time.elapsed().as_secs_f32();
        let progress = (elapsed / self.transition_duration).min(1.0);

        if progress >= 1.0 {
            self.transitioning = false;
            return self.env_manager.current().image.clone();
        }

        let w = self.env_manager.width;
        let h = self.env_manager.height;

        // Ease-out cubic: 1.0 - (1.0 - progress)^3
        let ease = 1.0 - (1.0 - progress).powi(3);
        let shift = ((ease * w as f32) as u32).min(w);

        let mut canvas = ImageBuffer::new(w, h);

        if self.transition_direction > 0 {
            // Slide right: old exits right, new enters from left
            for y in 0..h {
                // New entering on left: [0..shift] from to_image[w - shift..w]
                for x in 0..shift {
                    let src_x = w - shift + x;
                    canvas.put_pixel(x, y, *self.to_image.get_pixel(src_x, y));
                }
                // Old exiting on right: [shift..w] from from_image[0..w - shift]
                for x in shift..w {
                    let src_x = x - shift;
                    canvas.put_pixel(x, y, *self.from_image.get_pixel(src_x, y));
                }
            }
        } else {
            // Slide left: old exits left, new enters from right
            for y in 0..h {
                // Old exiting on left: [0..w - shift] from from_image[shift..w]
                for x in 0..(w - shift) {
                    let src_x = shift + x;
                    canvas.put_pixel(x, y, *self.from_image.get_pixel(src_x, y));
                }
                // New entering on right: [w - shift..w] from to_image[0..shift]
                for x in (w - shift)..w {
                    let src_x = x - (w - shift);
                    canvas.put_pixel(x, y, *self.to_image.get_pixel(src_x, y));
                }
            }
        }

        canvas
    }

    pub fn current_environment(&self) -> &Environment {
        self.env_manager.current()
    }

    pub fn is_original_room(&self) -> bool {
        self.env_manager.is_original_room() && !self.transitioning
    }
}
